### DISCUSSION TITLE
How can I use `tree-sitter test` with an external scanner not written in C/C++?

### DISCUSSION THREAD

**Discussion by CAD97** [tree-sitter/tree-sitter#1915] [Q&A] [answered]
I'm writing a parser with a scanner written in Rust. From the Rust bindings, everything is working to link.

`tree-sitter test`, however, fails to link, because it's not linking in the scanner code. I've not told the node bindings anything about the Rust implemented scanner, so this isn't too surprising... but how do I do so?

The `binding.gyp` has a clear place to add new C/C++ files for a C/C++ scanner, but is it possible to get Rust linked into the node bindings?

I don't really need/intend this parser to be usable from node (but it'd be nice for sure), but I do want to be able to use `tree-sitter test`.

**Comment by ahelwer** [comment, marked as answer]
Not answering your question, but do note every project that currently consumes tree-sitter uses a C/C++ toolchain to build the parser so writing your scanner in rust bars you from integrating with them. Things like neovim and such.

**Comment by ruifengx** [comment]
There is a `link_settings` section available in `binding.gyp`, and we can add libraries to link here. (See for example [here](https://gyp.gsrc.io/docs/UserDocumentation.md#support-for-mac-os-x-bundles) for the syntax.) I also found [this repository](https://github.com/wtfil/rust-in-node) which specifically deals with linking to Rust in a node module as a dynamic library.

I think we can do the following:
- Create a Rust scanner crate and build as `cdylib`;
- Specify the dependency in `binding.gyp` as shown [here](https://github.com/wtfil/rust-in-node/blob/master/src/native-cpp-ffi/binding.gyp);

Haven't tested it yet, but I presume it should work (though a bit ugly IMO). The Rust scanner should be built as `cdylib` rather than a `staticlib` because I believe we will need to manually link `libstd` ourselves otherwise. However this is only the case when linking for node; when using the parser from Rust we should be able to link as `rlib` as usual.

---

EDIT: it seems `libstd` _is_ included in `staticlib`s, so it might also just work with `staticlib` (and if it does I think this is preferable).

---

Actually this does not work. I checked the implementation of the CLI just now, and it just hard-coded the path `scanner.c`/`scanner.cc` in the `loader`. The `gyp` file is only used for building the binding for `node`. See the following line for details:
https://github.com/tree-sitter/tree-sitter/blob/b268e412ad4848380166af153300464e5a1cf83f/cli/loader/src/lib.rs#L314

We will need a PR targetting this part to support non C/C++ scanners in the `test` command.

**Comment by maxbrunsfeld** [comment]
I'd love to make it possible for `tree-sitter test` and `tree-sitter parse` to compile and link external scanners written in Rust. If someone wants to open a PR, I'd be open to it.

**Reply by ahelwer** [reply]
After fixing [a ridiculous bug](https://github.com/tlaplus-community/tree-sitter-tlaplus/pull/78) in my C++ external scanner I am now dying to RIIR lol
